### PR TITLE
Support CUDA 11.4 and support `compute_86`

### PR DESCRIPTION
**Pull request by takagi**
Blocked by https://github.com/chainer/chainer-test/pull/673.

Includes the following updates for CUDA 11.4:
- Compute capability support
- bundled FP16 header files
- library installer

### CODE REVIEW

**Comment by kmaehashi** [comment]
Hmm, seems NVRTC docs is outdated.

CUDA 11.2 / 11.3
```
>>> cupy.cuda.nvrtc.getSupportedArchs()
(35, 37, 50, 52, 53, 60, 61, 62, 70, 72, 75, 80, 86)
```

CUDA 11.4
```
>>> cupy.cuda.nvrtc.getSupportedArchs()
(35, 37, 50, 52, 53, 60, 61, 62, 70, 72, 75, 80, 86, 87)
```

**Comment by takagi** [comment]
I also suspect nvcc's. NVCC accepts `compute_86/sm_86` from CUDA 11.1.

**Comment by kmaehashi** [comment]
> I also suspect nvcc's. NVCC accepts compute_86/sm_86 from CUDA 11.1.

I see... `compute_86` is used in the official CUDA v11.1 examples, so I think it's safe to assume it's supported since 11.1.
https://github.com/NVIDIA/cuda-samples/blob/v11.1/Samples/bf16TensorCoreGemm/bf16TensorCoreGemm_vs2019.vcxproj

File: cupyx/tools/install_library.py [hunk 1]
```diff
@@ -174,6 +182,9 @@ def _make_nccl_record(
     }
 
 
+_nccl_records.append(_make_nccl_record(
```
Comment (kmaehashi) [inline cupyx/tools/install_library.py:185]: Could you use 2.10.3?

File: cupy_setup_build.py [hunk 2]
```diff
@@ -1004,7 +1004,17 @@ def _nvcc_gencode_options(cuda_version):
...
+        if cuda_version >= 11040:
+            arch_list = ['compute_35',
+                         'compute_50',
+                         ('compute_60', 'sm_60'),
+                         ('compute_61', 'sm_61'),
+                         ('compute_70', 'sm_70'),
+                         ('compute_75', 'sm_75'),
+                         ('compute_80', 'sm_80'),
+                         ('compute_86', 'sm_86'),
+                         'compute_86']
```
Comment (kmaehashi) [inline cupy_setup_build.py:1016]: This entry is intended as a fallback for all `compute_8X`.
```suggestion
                         'compute_80']
```

File: cupy/cuda/compiler.py [hunk 3]
```diff
@@ -133,7 +133,7 @@ def _get_arch():
         # CUDA 10.1 / 10.2
         _nvrtc_max_compute_capability = '75'
     else:
-        # CUDA 11.0 / 11.1 / 11.2
+        # CUDA 11.0 / 11.1 / 11.2 / 11.3 / 11.4
         _nvrtc_max_compute_capability = '80'
```
Comment (kmaehashi) [inline cupy/cuda/compiler.py:137]: Could you allow 86 for CUDA 11.1+?

File: cupy_setup_build.py [hunk 4]
```diff
@@ -1004,7 +1004,17 @@ def _nvcc_gencode_options(cuda_version):
         #
         #   https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#options-for-steering-gpu-code-generation
 
-        if cuda_version >= 11000:
+        if cuda_version >= 11040:
```
Comment (kmaehashi) [inline cupy_setup_build.py:1007]:
```suggestion
        if cuda_version >= 11010:
```

**Review by kmaehashi — CHANGES_REQUESTED** [review]

**Comment by kmaehashi** [comment]
Let me test `pfn-public-ci/cupy.linux.cuda114`.
/test

File: cupy_setup_build.py (hunk 2 above)
Comment (takagi) [reply cupy_setup_build.py:1016]: I think GPU with cc=80 and cc=86 are covered by the two lines above, and the line is for forward compatibility for future cc>86. Also I think nVidia doesn’t insert a new cc number below the maximum one. Do I have some misunderstanding?

**Comment by takagi** [comment]
Ah, sorry, I commited a suggested change before the test run finishes, and it cleared the CI status on github. (The test itself keeps running? https://ci.preferred.jp/cupy.linux.cuda114/74868/)

File: cupy_setup_build.py (hunk 2 above)
Comment (kmaehashi) [reply cupy_setup_build.py:1016]: Ah, you are correct, that seems like a correct approach to provide forward compatibility 😄 
https://github.com/NVIDIA/cuda-samples/blob/v11.4/Samples/vectorAddMMAP/Makefile#L323-L328

File: cupyx/tools/install_library.py (hunk 1 above)
Comment (kmaehashi) [reply cupyx/tools/install_library.py:185]: Let's do this in separate NCCL 2.10 PR 😄

File: cupy/cuda/compiler.py (hunk 3 above)
Comment (takagi) [reply cupy/cuda/compiler.py:137]: Can we also add that of `87`? It seems that nvcc 11.4 also supports `compute_87/sm_87`.

File: cupy_setup_build.py [hunk 5]
```diff
@@ -1004,7 +1004,27 @@ def _nvcc_gencode_options(cuda_version):
...
+            arch_list = ['compute_50',
```
Comment (takagi) [inline cupy_setup_build.py:1008]: Kepler `compute_35` support is deprecated in CUDA 11.4.

File: cupy/cuda/compiler.py (hunk 3 above)
Comment (kmaehashi) [reply cupy/cuda/compiler.py:137]: I think it's safer to keep `86` here (when the new GPU with CC = 90 is released users may hit this) as it's undocumented and CUDA examples don't use it.
https://github.com/NVIDIA/cuda-samples/search?l=Makefile&q=SMS

File: cupy_setup_build.py [hunk 6]
```diff
@@ -1004,7 +1004,28 @@ def _nvcc_gencode_options(cuda_version):
...
+            arch_list = ['compute_35',
+                         'compute_50',
+                         ('compute_60', 'sm_60'),
+                         ('compute_61', 'sm_61'),
+                         ('compute_70', 'sm_70'),
+                         ('compute_75', 'sm_75'),
+                         ('compute_80', 'sm_80'),
+                         ('compute_86', 'sm_86'),
+                         ('compute_87', 'sm_87'),
+                         'compute_87']
```
Comment (kmaehashi) [inline cupy_setup_build.py:1017]:
```suggestion
                         'compute_86']
```

**Review by kmaehashi — CHANGES_REQUESTED** [review]

File: cupyx/tools/install_library.py (hunk 1 above)
Comment (takagi) [reply cupyx/tools/install_library.py:185]: Sure!

**Comment by takagi** [comment]
pfnCI, test this please.

**Review by kmaehashi — APPROVED** [review]

**Pull request merged**